src/ripple/overlay/MessageScheduler.md

```diff
@@ -0,0 +1,276 @@
+# MessageScheduler
+
+At times, our server needs to communicate with its peers:
+sending them transactions, proposals, and validations, or requesting data.
+At different times we may be sending more than we are receiving, and vice versa.
+We have to strike a balance between our ability to flood peers with
+notifications and requests and their ability to digest those notifications and
+respond to those requests.
+If not, then we run the risk of overwhelming our peers in activity
+indistinguishable from a denial-of-service attack.
+
+Each message to a peer effectively corresponds to a remote job executed on that
+peer.
+We already recognize that local threads are a limited resource,
+and that local jobs are a demand on that resource.
+We throttle the number of simultaneously executing local jobs to avoid
+overloading the underlying machine's capacity for simultaneously executing
+threads,
+and we schedule local jobs according to priority during periods of high
+contention.
+We should do the same for remote jobs, i.e. messages.
+
+
```

Comment:
Maybe explain the existing approach as a comparison to show the benefit of the MessageScheduler?

Reply:
Thanks!